crates/gen-spidermonkey/src/lib.rs

```diff
@@ -0,0 +1,2180 @@
+                                // []
+                            }
+                            // []
+                            self.inst(Instruction::LocalGet(1));
+                            // [i32]
+                            self.inst(Instruction::LocalGet(2));
+                            // [i32 i32]
+                            self.inst(Instruction::Call(
+                                self.gen.spidermonkey_import("SMW_finish_returns"),
+                            ));
+                            // []
+                        }
+
+                        // NB: only clear the JS operands after we've attached
+                        // the return value(s) to the `JS::CallArgs`.
+                        self.gen.clear_js_operands(self.blocks.last_mut().unwrap());
+
+                        // Return `true`, meaning that a JS exception was not thrown.
+                        //
+                        // []
+                        self.inst(Instruction::I32Const(1));
+                        // [i32]
+                        self.inst(Instruction::Return);
+                        // []
+                    }
+                    witx2::abi::LiftLower::LiftArgsLowerResults => {
+                        self.gen.clear_js_operands(self.blocks.last_mut().unwrap());
+
+                        // Get the return values out of their locals and push
+                        // them onto the Wasm stack.
+                        //
+                        // []
+                        for _ in 0..*amt {
+                            let local = pop_wasm(operands);
+                            self.inst(Instruction::LocalGet(local));
+                        }
+                        // [R...]
+                        self.inst(Instruction::Return);
+                        // []
+                    }
+                }
+            }
+            witx2::abi::Instruction::Witx { instr: _ } => {
+                unreachable!("we do not support the preview1 ABI")
+            }
+        }
+    }
+
+    fn allocate_typed_space(
+        &mut self,
+        _iface: &witx2::Interface,
+        _ty: witx2::TypeId,
+    ) -> Self::Operand {
+        todo!()
+    }
+
+    fn i64_return_pointer_area(&mut self, amt: usize) -> Self::Operand {
+        assert!(amt <= self.gen.i64_return_pointer_area_size);
+        let local = self.new_local(wasm_encoder::ValType::I32);
+
+        // []
+        self.inst(Instruction::GlobalGet(RET_PTR_GLOBAL));
+        // [i32]
+        self.inst(Instruction::LocalSet(local));
+        // []
+
+        Operand::Wasm(local)
+    }
+
+    fn push_block(&mut self) {
+        self.blocks.push(vec![]);
+    }
+
+    fn finish_block(&mut self, results: &mut Vec<Self::Operand>) {
+        self.block_results.push(results.to_vec());
+    }
+
+    fn sizes(&self) -> &witx2::SizeAlign {
+        todo!()
+    }
+
+    fn is_list_canonical(&self, _iface: &witx2::Interface, _ty: &witx2::Type) -> bool {
+        false
```

Comment:
Oh I see this might answer my question from earlier. Does this mean though that things like `list<u8>` have element-per-element processing rather than something that lowers down to a `memory.copy`? If so I think it might be good to have a `TODO` here to at least change that for `TypedArray`-style things.